### ISSUE TITLE
Accessing active datasource in onPositionDiscontinuity

### ISSUE THREAD

**Issue by Tolriq** [google/ExoPlayer#7525]
### [REQUIRED] Searched documentation and issues

I did :)

### [REQUIRED] Question
I'm having an hard time to get the active datasource when the media change in a ```ConcatenatingMediaSource```.

I use a specific ```DataSourceFactory``` that extends ```HttpDataSource.BaseFactory``` and prepare the media in the ```open(dataSpec)```.
During this preparation the media can be transcoded or not and depending on many factor the resulting may support some things like seeking or not.
I need to be able to access this information to do things like not calling normal seek but restart the transcoding at the new position.

The problem is that in ```ConcatenatingMediaSource``` ```open``` can be called to prepare the next media while the active media is still not closed.

So the question is how can I access the datasource from the player ```onPositionDiscontinuity``` callback. Or how can I access the ```MediaSource``` from the ```DataSource``` to be able to modify the tag?

**Comment by AquilesCanta** [comment]
Why don't you assign a specific `DataSourceFactory` to each `MediaSource` so that each `DataSourceFactory` knows to which MediaSource they are assigned?

So when the media source calls `open(...)` for preparation (or whatever), you know which is the calling MediaSource, since the `DataSourceFactory` knows the corresponding `MediaSource`.

**Comment by Tolriq** [comment]
Isn't that highly inefficient to create a DefaultDataSourceFactory and my factory for every single items for very large playlists?

**Comment by tonihei** [comment]
`DefaultDataSourceFactory` is just a thin wrapper creating other `DataSource` instances, so there should be no issue with creating multiple factories if needed.

Not sure about your exact use case, but you can also consider one of the following:
 - Use `ResolvingDataSource` to add additional loading steps based on the `DataSpec`. You could detect whatever you need from the `DataSpec` URL, custom header key-value pairs or from using the `customData` object. Whether this is easily possible depends on the details of what you need and what you are trying to achieve.
 - Set a `tag` on each `MediaSource` (or `MediaItem`) that can be retrieved by `Player.getCurrentTag()` in `onPositionDiscontinuity`.

**Comment by Tolriq** [comment]
@tonihei  Ok so let me try to better explain the need that is.

The player can play media from many different sources and some of them can transcode the media on the fly.
The transcoding or not decision is made at the ```DataSource.open``` to avoid wasting server resources.
When transcoding some server for some media type and transcoded media can generate non seekable media via normal player seek functions.
I can detect and know this only when transcoding start so in ```DataSource.open```

But I need to know this at the player / UI side to either disable the seek functions or intercept the seek to be actual restart of the media at the proper new starting point.

The problem is that with gapless / ConcateningMediaSource the ```DataSource.open``` is called independently of the active media.

And from ```DataSource.open``` I can't access the corresponding MediaItem/MediaSource to update the tag that I could then use in ```onPositionDiscontinuity```

So TL;DR I need to access the MediaSource/MediaItem from the ```DataSource.open``` not sure ```ResolvingDataSource``` brings anything here as it's before the open.

**Comment by tonihei** [comment]
`ResolvingDataSource` is essentially just an easy wrapper to intercept the `DataSpec` to the `DataSource`. If the actual information you need is only available afterwards, you would probably need to follow @AquilesCanta's advice and use a `DataSource.Factory` for each of your `DataSources` that has access to your tag object:

```
MediaSource mediaSource = 
    new ProgressiveMediaSource.Factory(() -> new MyCustomDataSource(tag))
      .setTag(tag)
      .createMediaSource(MediaItem.forUri(uri));
```

Aside: If you make the `tag` object mutable be aware that the loading takes place on another thread, so you need additional thread safety around the updates.

**Comment by Tolriq** [comment]
Thanks that's what I do, but the new playlist api is all about MediaItem directly no? Or I can use MediaSources directly and still use the new playlist api instead of concateningmediasource?

**Comment by tonihei** [comment]
Yes, you can. The `ExoPlayer` interface defines all relevant methods for that (e.g. `setMediaSource`, `addMediaSource` etc)

**Comment by Tolriq** [comment]
Ok thanks, closing this one too as there's no other way and factory are small :)
